lib/site_prism/element_container.rb

```diff
@@ -228,14 +228,37 @@ def deduce_iframe_element_find_args(args)
     end
 
     def extract_section_options(args, &block)
-      if args.first.is_a?(Class)
-        section_class = args.shift
-      elsif block_given?
-        section_class = Class.new(SitePrism::Section, &block)
+      section_class = deduce_section_class(args, &block)
+      arguments = deduce_search_arguments(section_class, args)
+      [section_class, arguments]
+    end
+
+    def deduce_section_class(args, &block)
+      if args.first.is_a?(Class) && args.first.ancestors.include?(SitePrism::Section)
+        klass = base_class = args.shift
+      end
+
+      if block_given?
+        klass = Class.new(base_class || SitePrism::Section, &block)
+      end
+
+      unless klass
+        raise ArgumentError, "You should provide section class either \
+as a block, or as the second argument."
+      end
+      klass
+    end
+
+    def deduce_search_arguments(section_class, args)
+      if args && !args.empty?
```

Comment:
for both this call and the one below we've got a double chain just checking for their presence safety before evaluating their length.

Could we not maybe create some sort of private method like `present?` like I've done in another PR that takes an array input. Or is that over-analysing do you think?